pkg/services/ngalert/eval/eval.go

```diff
@@ -24,28 +25,72 @@ import (
 
 var logger = log.New("ngalert.eval")
 
-//go:generate mockery --name Evaluator --structname FakeEvaluator --inpackage --filename evaluator_mock.go --with-expecter
-type Evaluator interface {
-	// ConditionEval executes conditions and evaluates the result.
-	ConditionEval(ctx EvaluationContext, condition models.Condition) Results
-	// QueriesAndExpressionsEval executes queries and expressions and returns the result.
-	QueriesAndExpressionsEval(ctx EvaluationContext, data []models.AlertQuery) (*backend.QueryDataResponse, error)
+type EvaluatorFactory interface {
 	// Validate validates that the condition is correct. Returns nil if the condition is correct. Otherwise, error that describes the failure
 	Validate(ctx EvaluationContext, condition models.Condition) error
+	// BuildRuleEvaluator build an evaluator pipeline ready to evaluate a rule's query
+	Create(ctx EvaluationContext, condition models.Condition) (ConditionEvaluator, error)
+}
+
+//go:generate mockery --name ConditionEvaluator --structname ConditionEvaluatorMock --with-expecter --output eval_mocks --outpkg eval_mocks
+type ConditionEvaluator interface {
+	// EvaluateRaw evaluates the condition and returns raw backend response backend.QueryDataResponse
+	EvaluateRaw(ctx context.Context, now time.Time) (resp *backend.QueryDataResponse, err error)
+	// Evaluate evaluates the condition and converts the response to Results
+	Evaluate(ctx context.Context, now time.Time) (Results, error)
+}
+
+type conditionEvaluator struct {
+	pipeline          expr.DataPipeline
+	expressionService *expr.Service
+	condition         models.Condition
+	evalTimeout       time.Duration
+}
+
+func (r *conditionEvaluator) EvaluateRaw(ctx context.Context, now time.Time) (resp *backend.QueryDataResponse, err error) {
+	defer func() {
+		if e := recover(); e != nil {
```

Comment:
How common do you expect panics to be? Is it something an end-user can trigger?

I'm not a big fan of recovering panics, especially if we don't have a high-visibility way to know about and fix them.

Reply:
I think this is copied from the existing code

Reply:
I think this is copied from the present code

Reply:
George is correct. My guess is that the panic can come from the plugin